vortex-array/src/compute/min_max.rs

```diff
@@ -0,0 +1,114 @@
+use vortex_dtype::Nullability::NonNullable;
+use vortex_error::{vortex_err, VortexError, VortexResult};
+use vortex_scalar::Scalar;
+
+use crate::stats::{Precision, Stat, Statistics};
+use crate::{Array, Encoding, IntoCanonical};
+
+pub type MinMaxResult = (Option<Scalar>, Option<Scalar>);
+
+/// Computes the min and max of an array, returning the (min, max) values
+pub trait MinMaxFn<A> {
+    fn min_max(&self, array: &A) -> VortexResult<MinMaxResult>;
+}
+
+impl<E: Encoding> MinMaxFn<Array> for E
+where
+    E: MinMaxFn<E::Array>,
+    for<'a> &'a E::Array: TryFrom<&'a Array, Error = VortexError>,
+{
+    fn min_max(&self, array: &Array) -> VortexResult<MinMaxResult> {
+        let (array_ref, encoding) = array.try_downcast_ref::<E>()?;
+        MinMaxFn::min_max(encoding, array_ref)
+    }
+}
+
+pub fn min_max(array: impl AsRef<Array>) -> VortexResult<MinMaxResult> {
+    let array = array.as_ref();
+
+    let min_max = if let Some(fn_) = array.vtable().min_max_fn() {
+        fn_.min_max(array)?
+    } else {
+        println!("NotImplemented: min_max {:?}, ", array.encoding());
+        let canonical = array.clone().into_canonical()?;
+        println!(
+            "NotImplemented, into_canonical: min_max {:?}, try",
+            canonical.encoding()
+        );
+        if let Some(fn_) = canonical.vtable().min_max_fn() {
+            fn_.min_max(canonical.as_ref())?
+        } else {
+            return Err(vortex_err!(NotImplemented: "min_max", array.encoding()));
+        }
+    };
+
+    if let (Some(min), _) = &min_max {
+        debug_assert_eq!(
+            min.dtype(),
+            &array.dtype().with_nullability(NonNullable),
+            "MinMax min dtype mismatch {}",
+            array.encoding()
+        );
+
+        array.set(Stat::Min, Precision::exact(min.clone().into_value()));
+    }
+
+    if let (_, Some(max)) = &min_max {
+        debug_assert_eq!(
+            max.dtype(),
+            &array.dtype().with_nullability(NonNullable),
+            "MinMax min dtype mismatch {}",
+            array.encoding()
+        );
+        array.set(Stat::Max, Precision::exact(max.clone().into_value()));
```

Comment:
This is a no-op for a ViewedArray, so i think your compute_statistics function cannot assume you can propagate these via the stats set

Reply:
Yep, that is annoying